apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -1,267 +1,394 @@
-      case 'paper_parked':
-        // Nothing to do - frontend triggers status change. This will change soon so not
-        // worth centralizing machine logic
-        break;
-      case 'printing_ballot':
-        if (!isPaperInScanner(paperHandlerStatus)) {
-          this.status = 'ballot_printed';
-        }
-        break;
-      case 'ejecting':
-        if (!isPaperAnywhere(paperHandlerStatus)) {
-          this.status = 'no_paper';
-        }
-        break;
-      case 'ballot_printed':
-        // Unimplemented
-        break;
+    switch (true) {
+      case state.matches('no_paper'):
+        return 'no_paper';
+      case state.matches('loading_paper'):
+        return 'loading_paper';
+      case state.matches('waiting_for_ballot_data'):
+        return 'waiting_for_ballot_data';
+      case state.matches('printing_ballot'):
+        return 'printing_ballot';
+      case state.matches('scanning'):
+        return 'scanning';
+      case state.matches('interpreting'):
+        return 'interpreting';
+      case state.matches('presenting_ballot'):
+        return 'presenting_ballot';
       default:
-        throwIllegalValue(this.status);
+        return 'no_hardware';
     }
-    return this.status;
   }
 
-  resetStatus(): void {
-    this.status = 'no_paper';
-  }
+  printBallot(
+    pdfData: Uint8Array,
+    options: Partial<ImageConversionOptions> = {}
+  ): Promise<void> {
+    this.machineService.send({
+      type: 'VOTER_INITIATED_PRINT',
+    });
 
-  async logStatus(): Promise<void> {
-    debug('%O', await this.driver.getPaperHandlerStatus());
+    return driverPrintBallot(this.driver, pdfData, options);
   }
+}
 
-  /**
-   * Parks paper inside the handler. If there is no paper to park, returns
-   * negative acknowledgement. If paper already parked, does nothing and returns
-   * positive acknowledgement. When parked, parkSensor should be true.
-   */
-  async parkPaper(): Promise<void> {
-    debug('+parkPaper');
-    this.status = 'parking_paper';
-    this.lastActionInitiatedTime = Date.now();
-    await this.driver.parkPaper();
-    debug('-parkPaper');
-  }
+function paperHandlerStatusToEvent(
+  paperHandlerStatus: PaperHandlerStatus
+): PaperHandlerStatusEvent {
+  let event: PaperHandlerStatusEvent = { type: 'NO_PAPER_IN_FRONT' };
 
-  /**
-   * Moves paper to the front for voter to see, but hangs on to the paper.
-   * Equivalent to "reject hold." How do we differentiate the present paper
-   * state from the state where paper has not been picked up yet?
-   */
-  async presentPaper(): Promise<void> {
-    await this.driver.presentPaper();
+  if (isPaperInScanner(paperHandlerStatus)) {
+    if (isPaperInOutput(paperHandlerStatus)) {
+      event = { type: 'BALLOT_PRINTED' };
+    } else {
+      event = { type: 'PAPER_INSIDE' };
+    }
+  } else if (isPaperReadyToLoad(paperHandlerStatus)) {
+    event = { type: 'PAPER_READY_TO_LOAD' };
   }
+  debug(`Emitting event ${event.type}`);
+  return event;
```

Comment:
I would recommend having the scanner status events all be descriptive of what the scanner sees rather than interpreting them. E.g. if paper is in the output position, it doesn't necessarily mean there was a ballot printed - that's contextual information that you can only know based on when you receive the status. So I might change that to `PAPER_IN_OUTPUT` or something. That way if you get a scanner status that says `PAPER_IN_OUTPUT` in a time other than when the ballot was printed, you know something went wrong.